### ISSUE TITLE
Using spans instead of raw pointers in the API

### ISSUE THREAD

**Issue by fsimonis** [precice/precice#1577]
**Please describe the problem you are trying to solve.**

The API of preCICE is simple, but extremely easy to misuse.
All chunks of data such as IDs, coordinates, or values are modelled using raw pointers.
preCICE infers many sizes from a single size attribute.
This makes the proposed #1559 rather dangerous as it is not clear if the user has reserved enough memory for scalar vs vector data.

**Describe the solution you propose.**

Migrate all raw pointers to span types.

**Examples**

```cpp
  void setMeshVertices(
      int                     meshID,
      std::span<const double> positions,
      std::span<int>          ids);

  void writeBlockVectorData(
      int                     dataID,
      std::span<const int>    valueIndices,
      std::span<const double> values);

  void setMeshTriangles(
      int                  meshID,
      std::span<const int> vertices);

  void getMeshVerticesAndIDs(
      const int         meshID,
      std::span<int>    ids,
      std::span<double> coordinates) const;
```

Taking #1461 into account
```cpp
  // participant.setMeshVertices("Mesh", coordinates, vertices);
  void setMeshVertices(
      std::string_view        mesh,
      std::span<const double> positions,
      std::span<int>          ids);

  // participant.writeBlockVectorData("Mesh", "Data", vertices, values);
  void writeBlockVectorData(
      std::string_view        mesh,
      std::string_view        data,
      std::span<const int>    valueIndices,
      std::span<const double> values);

  // participant.setMeshTriangles("Mesh", trianglePoints);
  void setMeshTriangles(
      std::string_view     mesh,
      std::span<const int> vertices);

  // participant.getMeshVerticesAndIDs("Mesh", ids, coords);
  void getMeshVerticesAndIDs(
      std::string_view   mesh,
      std::span<int>     ids,
      std::span<double>  coordinates) const;
```

Examples for calling:
```cpp
participant.setMeshTriangles("Mesh", {pointer, size});                  // pointer + size
participant.setMeshTriangles("Mesh", myvector);                         // std::vector
participant.setMeshTriangles("Mesh", {myvector.begin(), someiterator}); // contigous iterators
participant.setMeshTriangles("Mesh", {{1,2,3,4}});                      // initializer list
```

**Upsides:**
* preCICE can check if input chunks are correctly sized. Example 15 coordinate values given, but the mesh is 2d.
* preCICE can check if output chunks are big enough. Example: 10 2d coordinates -> 5 vertices, but output fits only 4.
* Direct conversion from `std::vector`, and iterators to contiguous data such `std::vector::iterator` or `double *`.
* This would align the C++ interface more with the python interface, which is the most used interface. The rust interface already uses slices, which are comparable to spans.
* This would make #1559 safe

**Downsides:**
* Even more changes to the API (realistically our users have to touch these functions after #1461 anyhow)
* Inconsistent with C and Fortran bindings
* Fortran bindings will become even more inconsistent
* C-bindings could use a span struct, but this will become verbose.
* We require a dependency for the interface (but we ship it as well)

**Open questions:**
* How do the C-bindings change? Leave it? Span struct? One size per pointer may lead to confusion in case of passing coordinates or so.
* Same for Fortran.

**Technicalities:**
* `std::span` is part of C++20, which requires GCC 10, which is available in Ubuntu 22.04 LTS, not in Ubuntu 20.04 LTS
* In the meantime, we could use our internally used header-only span type and install it alongside the API headers. Once Ubuntu 24.04 LTS is released, we move to C++20, use `std::span` and remove the header.

**Describe alternatives you've considered**
Wait for C++20 (April 2024), realistically this would be breaking though.

**Additional context**

This is extrapolated feedback from the preCICE Workshop 2023. _**You** should join next time_ :wink:

**Comment by fsimonis** [comment]
A follow-up regarding the C bindings which can be implemented in various ways

The cpp interface of 
```cpp
void setMeshVertices(
      std::string_view        mesh,
      std::span<const double> positions,
      std::span<int>          ids);
```

Version 1: The user is forces to provide the sizes of passed arrays
```c
void precicec_setMeshVertices(
      const char   *meshName,
      const double *positions,
      int           positionsSize,
      int          *ids,
      int           idsSize
){
 impl->setMeshVertices(
    meshName,
    {positions, positionsSize},
    {ids, idsSize});
}
```

Version 2: Unchanged API, preCICE determines the size of positions itself, which is unsafe.
```c
void precicec_setMeshVertices(
      const char   *meshName,
      int           size,
      const double *positions,
      int          *ids,
){
 auto positionsSize = size * impl->getDimensions();
 impl->setMeshVertices(
    meshName,
    {positions, positionsSize},
    {ids, size});
}
```

Version 3: We require spans
```c
typedef struct {
    int *data;
    int  size;
} precicec_int_span;

typedef struct {
    const double *data;
    int           size;
} precicec_cdouble_span;

void precicec_setMeshVertices(
      const char           *meshName,
      precicec_cdouble_span positions,
      precicec_int_span     ids
){
 impl->setMeshVertices(meshName,positions,ids);
}

void calling_3d_code() {
  int * ids;
  double* pos;
  int id_count;
  id_count = ...;
  ids = malloc(id_count * sizeof(double));
  pos = malloc(id_count * 3 * sizeof(double)); 
  precicec_setMeshVertices("MeshName",
    precicec_int_span{ids, id_count},
    precicec_cdouble_span{pos, id_count*3});
}
```

**Comment by uekerman** [comment]
Thanks for the additional examples!

For me, the main advantage of `std::span` is in making our C++ API more C++-like, i.e. we can now directly pass `std::vector`.

For the C bindings (and probably similarly for Fortran), I am not yet too convinced of this change. Is "Version 1" really that much safer than "Version 2" to justify the change? In all three versions, we trust the user to pass the correct size values (in contrast to the C++ version where size is deduced automatically from e.g. a `std::vector`).
I would vote for "Version 2".

**Comment by davidscn** [comment]
Discussing in person. Conclusion:

- we want to use and install our own `precice::span` and make it accessible for the C++ API
- we want to keep the C and Fortran bindings as they are right now, i.e., go with version 2 proposed above.

**Comment by fsimonis** [comment]
Closed via #1641 #1638 #1626
